test/integration/etcd_storage_path_test.go

```diff
@@ -0,0 +1,1079 @@
+			all := &[]cleanupData{}
+			defer func() {
+				if !t.Failed() { // do not cleanup if test has already failed since we may need things in the etcd dump
+					if err := client.cleanup(all); err != nil {
+						t.Fatalf("failed to clean up etcd: %#v", err)
+					}
+				}
+			}()
+
+			if err := client.createPrerequisites(mapper, testNamespace, testData.prerequisites, all); err != nil {
+				t.Errorf("failed to create prerequisites for %s from %s: %#v", kind, pkgPath, err)
+				return
+			}
+
+			if shouldCreate { // do not try to create items with no stub
+				if err := client.create(testData.stub, testNamespace, mapping, all); err != nil {
+					t.Errorf("failed to create stub for %s from %s: %#v", kind, pkgPath, err)
+					return
+				}
+			}
+
+			output, err := getFromEtcd(keys, testData.expectedEtcdPath)
+			if err != nil {
+				t.Errorf("failed to get from etcd for %s from %s: %#v", kind, pkgPath, err)
+				return
+			}
+
+			expectedGVK := gvk
+			if testData.expectedGVK != nil {
+				expectedGVK = *testData.expectedGVK
+			}
+
+			actualGVK := getGVK(output)
+			if actualGVK != expectedGVK {
+				t.Errorf("GVK for %s from %s does not match, expected %s got %s", kind, pkgPath, expectedGVK, actualGVK)
+			}
+
+			if !deepDerivativeMap(input, output) {
+				t.Errorf("Test stub for %s from %s does not match: %s", kind, pkgPath, diff.ObjectGoPrintDiff(input, output))
+			}
+		}()
+	}
+
+	if inEtcdData, inEtcdSeen := diffMaps(etcdStorageData, etcdSeen); len(inEtcdData) != 0 || len(inEtcdSeen) != 0 {
+		t.Errorf("etcd data does not match the types we saw:\nin etcd data but not seen:\n%s\nseen but not in etcd data:\n%s", inEtcdData, inEtcdSeen)
+	}
+
+	if inEphemeralWhiteList, inEphemeralSeen := diffMaps(ephemeralWhiteList, ephemeralSeen); len(inEphemeralWhiteList) != 0 || len(inEphemeralSeen) != 0 {
+		t.Errorf("ephemeral whitelist does not match the types we saw:\nin ephemeral whitelist but not seen:\n%s\nseen but not in ephemeral whitelist:\n%s", inEphemeralWhiteList, inEphemeralSeen)
+	}
+
+	if inKindData, inKindSeen := diffMaps(kindWhiteList, kindSeen); len(inKindData) != 0 || len(inKindSeen) != 0 {
+		t.Errorf("kind whitelist data does not match the types we saw:\nin kind whitelist but not seen:\n%s\nseen but not in kind whitelist:\n%s", inKindData, inKindSeen)
+	}
+}
+
+type prerequisite struct {
+	gvrData unversioned.GroupVersionResource
+	stub    string
+}
+
+type empty struct{}
+
+type cleanupData struct {
+	obj     runtime.Object
+	mapping *meta.RESTMapping
+}
+
+func gvr(g, v, r string) unversioned.GroupVersionResource {
+	return unversioned.GroupVersionResource{Group: g, Version: v, Resource: r}
+}
+
+func getGVK(obj map[string]interface{}) unversioned.GroupVersionKind {
+	return unversioned.FromAPIVersionAndKind(obj["apiVersion"].(string), obj["kind"].(string))
+}
+
+var requiredFields = sets.NewString("name", "namespace")
+
+func deepDerivativeMap(a, b map[string]interface{}) bool {
```

Comment:
this seems way too complicated just to check metadata.name and metadata.namespace